File: typegate/tests/runtimes/graphql/typegraphs/python/graphql.py
from typegraph.providers.prisma import PrismaRuntime
from typegraph.runtimes.graphql import GraphQLRuntime

from typegraph import Graph, Policy, effects, t, typegraph

_user = t.struct(
    {"id": t.string(as_id=True), "name": t.string()},
    name="User",
)

create_user_inp = t.struct(
    {"name": t.string(), "username": t.string(), "email": t.email()}
)


@typegraph()
def graphql(g: Graph):
    # graphql = GraphQLRuntime("https://example.com/api/graphql")
    gql = GraphQLRuntime("https://graphqlzero.almansi.me/api")
    public = Policy.public()
    db = PrismaRuntime("graphql", "POSTGRES")

    message = t.struct(
        {
            "id": t.integer(as_id=True, config=["auto"]),
            "title": t.string(),
            # highlight-next-line
            "user_id": t.string(name="uid"),
            # highlight-next-line
            "user": gql.query(
                t.struct(
                    {
                        # highlight-next-line
                        "id": t.string(as_id=True).from_parent("uid")
                    }
                ),
                t.optional(_user),
            ),
        },
        name="message",
    )

    g.expose(
        users=gql.query(
            t.struct(),
            t.struct(
                {"data": t.list(_user)},
            ),
        ).with_policy(public),
        user=gql.query(
            t.struct({"id": t.string(as_id=True)}),
            _user,
        ).with_policy(public),
        create_message=db.create(message).with_policy(public),
        create_user=gql.mutation(
            t.struct(
                {
                    "input": create_user_inp,
                }
            ),
            _user,
            effects.create(False),
        ).with_policy(public),
        messages=db.find_many(message).with_policy(public),
    )

File: tests/runtimes/graphql/typegraphs/deno/graphql.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Policy, t, typegraph } from "@typegraph/sdk";
import { GraphQLRuntime } from "@typegraph/sdk/runtimes/graphql";
import * as effects from "@typegraph/sdk/effects";
import { PrismaRuntime } from "@typegraph/sdk/providers/prisma";

const user = t.struct(
  {
    id: t.string({}, { asId: true }),
    name: t.string(),
    // TODO more fields with more types
  },
  { name: "User" },
);

const createUserInput = t.struct({
  name: t.string(),
  username: t.string(),
  email: t.string(),
}, { name: "CreateUserInput" });

export const tg = await typegraph("graphql", (g) => {
  const graphql = new GraphQLRuntime("https://graphqlzero.almansi.me/api");
  const pub = Policy.public();
  const db = new PrismaRuntime("graphql", "POSTGRES");

  const message = t.struct(
    {
      id: t.integer({}, { asId: true, config: { auto: true } }),
      title: t.string(),
      user_id: t.string(),
      user: graphql.query(
        t.struct({
          id: t.string({}, { asId: true }).fromParent("user_id"),
        }),
        t.optional(user),
      ),
    },
    { name: "message" },
  );

  g.expose({
    user: graphql
      .query(t.struct({ id: t.string({}, { asId: true }) }), user)
      .withPolicy(pub),
    users: graphql
      .query(
        t.struct({}),
        t.struct({
          data: t.list(user),
        }),
      )
      .withPolicy(pub),
    createUser: graphql
      .mutation(
        t.struct({
          input: createUserInput,
        }),
        user,
        effects.create(false),
      )
      .withPolicy(pub),
    create_message: db.create(message).withPolicy(pub),
    messages: db.findMany(message).withPolicy(pub),
  });
});
